sql: Support RESET and SET TO DEFAULT

boxofrad opened the pull request:
Howdy! 👋 

This adds support for both `SET ... TO DEFAULT` (#13952) and `RESET ...` (#13948) statements, required to support the Flyway schema migration tool.

It's broken down into two commits:

- 7f866e4c4 - `varGen` refactoring (as [suggested by](https://github.com/cockroachdb/cockroach/issues/13952#issuecomment-285396237) @jordanlewis)
- 44a6510f8 - adds support for `TO DEFAULT` and the `RESET` keyword

This is my first contribution to CockroachDB, thanks for making me feel welcome so far! ❤️

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14105)
<!-- Reviewable:end -->

jordanlewis commented:
Hi @boxofrad, thanks very much for your contribution! We really appreciate it and one of us will review your PR soon. In the meantime it looks like your PR is failing a style check in CI - you can look at the failure in the build log in TeamCity, or just run `make check` and read the output.

jordanlewis commented:
This is really good work! I have a few minor suggestions for improvement

---

Reviewed 5 of 6 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/reset.go, line 31 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf71IrvRgHYUsabz9Oe:-Kf71IrvRgHYUsabz9Of:b-qey11k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ecafbb23cba562a202f3c24189f6adaf701ba/pkg/sql/reset.go#L31)):*
> ```Go
> 	if v, ok := varGen[name]; ok {
> 		if v.Reset == nil {
> 			return nil, fmt.Errorf("variable \"%s\" cannot be changed", name)
> ```

I think this should say `cannot be reset` instead of `cannot be changed` so the error message is a little more explicit about what went wrong.

---

*[pkg/sql/set.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf6ywV3TpoqzALxs7Lp:-Kf6ywV4xhrJbQKcYiE2:bobi4k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/set.go#L56)):*
> ```Go
> 	if v, ok := varGen[name]; ok {
> 		if v.Set == nil {
> 			return nil, fmt.Errorf("variable \"%s\" cannot be changed", name)
> ```

Can you please add a test for this error case in `testdata/set`?

---

*[pkg/sql/vars.go, line 30 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf7-CdxnrIJTara43BS:-Kf7-Ce-514kwhhlGb8c:b-717vr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/vars.go#L30)):*
> ```Go
> 	PgServerVersion = "9.5.0"
> )
> ```

This could use a comment, as could its constituent fields, whose inputs, outputs and side effects should be explained.

---

*[pkg/sql/parser/sql.y, line 1363 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf70mzVQzrbrGca6Uud:-Kf70mzWat6-xoEvSFuq:bslpmrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ecafbb23cba562a202f3c24189f6adaf701ba/pkg/sql/parser/sql.y#L1363)):*
> ```Yacc
>   }
> 
> // RESET DISTSQL
> ```

This comment should be `RESET name` to be more like the comments around it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-Kf728HyM_1v4cvPozjp:b-caqn9q)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Thanks for the feedback @jordanlewis!
I'm on vacation & will be away from a go compiler for the next few days :wink:, but I'll hopefully get those points addressed when I'm back on Thursday.

tamird commented:
Reviewed 5 of 6 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 165 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf77ttPm4xi-IU5yZ3J:-Kf77ttPm4xi-IU5yZ3K:b-fbzkxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7a0728886bebbc1843e2afeb5b91acc5654948b/pkg/sql/session.go#L165)):*
> ```Go
> }
> 
> // sessionDefaults mirrors fields in Session, for restoring default
> ```

this mirrors `varGen` quite closely; is there a sensible way to avoid the duplication?

---

*[pkg/sql/set.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76JiIGzK7agqB7OMT:-Kf76JiIGzK7agqB7OMU:bc52zda) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/set.go#L64)):*
> ```Go
> 	} else {
> 		switch name {
> 		// These settings are sent by various client drivers. We don't support
> ```

why leave these here? seems like you could move these to `varGen` with `nil` `Set` functions and remove this comment.

---

*[pkg/sql/show.go, line 143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76WXkrOt8N1B8tZtV:-Kf76WXlxKTLXhDnr86-:b-8icn3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/show.go#L143)):*
> ```Go
> func (p *planner) Show(n *parser.Show) (planNode, error) {
> 	origName := n.Name
> 	name := strings.ToUpper(n.Name)
> ```

why change the casing here? where before this was allocating a single string, it now allocates two (one for the up-case here, and another for the down-case below).

---

*[pkg/sql/vars.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76zIynZpbILyI1EWl:-Kf76zIynZpbILyI1EWm:be0gyst) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/vars.go#L189)):*
> ```Go
> 			return nil
> 		},
> 		Get: func(_ *planner) string {
> ```

nit (I realize this isn't your code): this can be `func(*planner) string` (the underscore is superfluous when all arguments are unbound).

applies to MAP_INDEX_KEYS and SERVER_VERSION as well.

---

*[pkg/sql/vars.go, line 241 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76qewqJRAccEN1VM2:-Kf76qewqJRAccEN1VM3:b-q58llt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/vars.go#L241)):*
> ```Go
> }
> 
> var varNames = func() []string {
> ```

nit: while you're here, this function only has one caller, so it might make sense to inline this implementation into that caller.

---

*[pkg/sql/vars.go, line 213 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf77VSiBZjZpU4xLzTa:-Kf77VSjddLanH3N3q3d:bss949a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7a0728886bebbc1843e2afeb5b91acc5654948b/pkg/sql/vars.go#L213)):*
> ```Go
> 			return "on"
> 		},
> 		Reset: func(_ *planner) error {
> ```

(second commit): can be `func(*planner) string`

---

*[pkg/sql/testdata/reset, line 5 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf78GOLQrrmWSUsMzAW:-Kf78GOMvkgUKgWwCB5V:b-rpqfyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ecafbb23cba562a202f3c24189f6adaf701ba/pkg/sql/testdata/reset#L5)):*
> ```
> 
> statement ok
> SET SEARCH_PATH = foo
> ```

please add some lowercase tests

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 5 of 6 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 167 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf7EOjw1R7d-Me9Nagv:-Kf7EOjxDGIKUKvKQY_G:b-yrqal9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7a0728886bebbc1843e2afeb5b91acc5654948b/pkg/sql/session.go#L167)):*
> ```Go
> // sessionDefaults mirrors fields in Session, for restoring default
> // configuration values in SET ... TO DEFAULT statements.
> type sessionDefaults struct {
> ```

Only `applicationName` and `database` belong here. All the other have static (i.e. global) defaults and therefore the default values do not deserve to be copied onto every session.

---

*[pkg/sql/set.go, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf7DOCJM2wE6BpYSyAi:-Kf7DOCJM2wE6BpYSyAj:b9fvctg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/set.go#L54)):*
> ```Go
> 	}
> 
> 	if v, ok := varGen[name]; ok {
> ```

Like Tamir says below, simply put all the things in `varGen` but with a no-op `Set` method and treat the case `!ok` with the error message "variable does not exist".

---

*[pkg/sql/set.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76JiIGzK7agqB7OMT:-Kf7DaPdiRpxI7WKOhDB:b-anz2se) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/set.go#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why leave these here? seems like you could move these to `varGen` with `nil` `Set` functions and remove this comment.
</blockquote></details>

I think they should be no-op `Set` methods instead of `nil` (we accept the SET but don't do anything). But yes, unify is good :)

---

*[pkg/sql/vars.go, line 209 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf7DzCm3RKUChF68BjB:-Kf7DzCnPmNbtlY9q0tD:bvi1fjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/vars.go#L209)):*
> ```Go
> 		},
> 	},
> 	`TIME ZONE`: {
> ```

Unless `make check` disapproves I would put the entire definition of this one and those that follow on a single line. I believe it fits and would make
the whole slightly more readable.

---

*[pkg/sql/parser/reset.go, line 23 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf7FiF9sz8mV6tNxb0w:-Kf7FiFA2qcn3va25pnB:bs2wm1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ecafbb23cba562a202f3c24189f6adaf701ba/pkg/sql/parser/reset.go#L23)):*
> ```Go
> 
> // Reset represents a RESET statement.
> type Reset struct {
> ```

No need to define a separate syntax node for this. Simply reuse `Set` and put a flag in there.

---

*[pkg/sql/parser/sql.y, line 1365 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf7FvVfEmB6hw8cv-yR:-Kf7FvVfEmB6hw8cv-yS:b-svzvdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a02ecafbb23cba562a202f3c24189f6adaf701ba/pkg/sql/parser/sql.y#L1365)):*
> ```Yacc
> // RESET DISTSQL
> reset_stmt:
>   RESET var_name
> ```

Add a test in `parse_test.go` for this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/set.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76JiIGzK7agqB7OMT:-KfQN4dMplD68ajHCRA1:bxsocu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/set.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think they should be no-op `Set` methods instead of `nil` (we accept the SET but don't do anything). But yes, unify is good :)
</blockquote></details>

The main reason was that everything in `varGen` appears in the `pg_catalog` table and when running `SHOW ALL`.

I guess a nice solution to that would be that we simply don't show anything with a `nil` `Set` method?

---

*[pkg/sql/show.go, line 143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76WXkrOt8N1B8tZtV:-KfQNZCtYmJm1gl5ev2s:b-bqo5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/show.go#L143)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why change the casing here? where before this was allocating a single string, it now allocates two (one for the up-case here, and another for the down-case below).
</blockquote></details>

It's just to maintain the existing behaviour around casing really. The `varGen` map now uses upper case keys, but `SHOW` has always shown the column name in lower-case.

Are we happy to forgo this behaviour?

---

*[pkg/sql/vars.go, line 241 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76qewqJRAccEN1VM2:-KfQOcy0wntGdVxCHMhN:bfon45u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/vars.go#L241)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: while you're here, this function only has one caller, so it might make sense to inline this implementation into that caller.
</blockquote></details>

Sorry, to clarify: did you mean only one referrer of `varNames`?

I found there's actually another one in `pg_catalog.go` :wink: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/set.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76JiIGzK7agqB7OMT:-KfRbN8TxJfgVQEJIprl:b3w4326) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/set.go#L64)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

The main reason was that everything in `varGen` appears in the `pg_catalog` table and when running `SHOW ALL`.

I guess a nice solution to that would be that we simply don't show anything with a `nil` `Set` method?
</blockquote></details>

I do not see why it is a problem to make them appear in pg_catalog. I think that's what we want.

---

*[pkg/sql/show.go, line 143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76WXkrOt8N1B8tZtV:-KfRbSwz7QnpufwJgASB:b-durjln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/show.go#L143)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

It's just to maintain the existing behaviour around casing really. The `varGen` map now uses upper case keys, but `SHOW` has always shown the column name in lower-case.

Are we happy to forgo this behaviour?
</blockquote></details>

Make them both lowercase.

---

*[pkg/sql/vars.go, line 241 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kf76qewqJRAccEN1VM2:-KfRb_MOm2GMGFkIT9VC:bq001om) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9145ce13c0005e984ef55b1ee06fb9926871edf/pkg/sql/vars.go#L241)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Sorry, to clarify: did you mean only one referrer of `varNames`?

I found there's actually another one in `pg_catalog.go` :wink: 
</blockquote></details>

Yes, varNames is indeed used in two places. Please keep it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

boxofrad committed:
sql: Unify variable getters and setters into a single map

Managing getters and setters in separate files often leads to new
variables being added that are either write-only or read-only.

This change also lays the ground work for adding proper support for
`SET ... TO DEFAULT` and `RESET ...` statements.

boxofrad committed:
sql: Support SET ... TO DEFAULT and RESET statements

This syntax is used by Flyway (popular schema migration tool).

boxofrad commented:
Hey Cockroach-ers! 🐛 

Thanks so much for taking the time to review my PR,
I've amended my commits with your feedback and rebased my branch on top of master so it should merge cleanly.

boxofrad commented:
Apologies for the failing build, looks like I lost the changes to `session.go` in the rebase.
Should be fixed now 😅

knz commented:
This is much better.
See my comment below about the Set logic.

Also, since you have decided to lowercase the name of the variables also in the error messages, please then extend the error message to indicate they come from the SET statement. For example instead of `fmt.Errorf("syntax: ...", ...)` write `fmt.Errorf("set syntax: ...", ...)`.

---

Reviewed 7 of 15 files at r5, 3 of 4 files at r6, 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/set.go, line 56 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kffkiojf-gUhFgMJe5F:-Kffkiojf-gUhFgMJe5G:ban1m7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3d6d8982201d63191e815711d7ccecf0a786c21/pkg/sql/set.go#L56)):*
> ```Go
> 
> 	if v, ok := varGen[name]; ok {
> 		if len(n.Values) == 0 {
> ```

This code is correct but I find the composition of conditionals a bit difficult to read.
We tend elsewhere to condition first on the type of syntax node, then do the corresponding semantic checks. Something like:
```go
v, ok := varGen[name]
if !ok {
   return nil, fmt.Errorf(...)
}
if len(n.Values) == 0 {
   n.SetMode = SetModeReset
}
switch n.SetMode {
case parser.SetModeAssign:
        if v.Set == nil {
             return nil, fmt.Errorf(...)
        }
        return &emptyNode{}, v.Set(...)
case parser.SetModeReset:
        if v.Reset == nil {
            return nil, fmt.Errorf(...)
        }
        return &emptyNode{}, v.Reset(...)
}
```

(you'd need to replace your newly created attribute `Set { Reset bool }` by a `Set { SetMode SetMode }` and create a new enumeration `SetMode` with values `SetModeAssign` and `SetModeReset`.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-Kffm_7tx9k-nHftHmFT:b-afhsc5)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Thanks @knz. Good point about the nested conditionals, it was starting to get a bit unwieldy!
Have applied that refactoring and fixed the error messages as requested.

tamird commented:
The last two commits seem to overlap quite a bit, so I'd probably squash them together if I were you.

---

Reviewed 3 of 15 files at r5, 10 of 10 files at r8, 1 of 1 files at r9, 4 of 4 files at r10, 7 of 7 files at r11.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-Kfg6bzMM5otRVrvtS1C:b-q17hpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
> ```Go
> 		}
> 	} else {
> 		return nil, fmt.Errorf("unknown variable: %q", name)
> ```

can you do the same origName thing here that is (was) done in show?

---

*[pkg/sql/set.go, line 61 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg8VojPzfARRr0llqm:-Kfg8VojPzfARRr0llqn:bgc5f6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b019d85e94e18454dd46db2a6bcd93e7636b7e7a/pkg/sql/set.go#L61)):*
> ```Go
> 
> 	if len(n.Values) == 0 {
> 		n.SetMode = parser.SetModeReset
> ```

can you use a local variable instead of mutating the AST?

---

*[pkg/sql/show.go, line 155 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6Yn0mUBx3Cvrc6Y3:-Kfg6Yn0mUBx3Cvrc6Y4:b7g0m3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/show.go#L155)):*
> ```Go
> 	default:
> 		if _, ok := varGen[name]; !ok {
> 			return nil, fmt.Errorf("unknown variable: %q", name)
> ```

why not use origName here?

---

*[pkg/sql/vars.go, line 38 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg5AGbQ1wMCy8YTdNZ:-Kfg5AGbQ1wMCy8YTdN_:bnx149k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/vars.go#L38)):*
> ```Go
> 	// Set performs mutations (usually on p.session) to effect the change
> 	// desired by SET commands.
> 	Set func(p *planner, ctx context.Context, values []parser.TypedExpr) error
> ```

`ctx` should be the first argument (I think we might even lint for this)

also (nit): this can be written without names: `Set func(context.Context, *planner, []parser.TypedExpr) error` -- ditto below.

---

*[pkg/sql/vars.go, line 223 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6ndclgUGydB8A1PD:-Kfg6nddpCKd-MzampCO:b-qex335) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/vars.go#L223)):*
> ```Go
> 		Get: func(p *planner) string { return p.txn.UserPriority().String() },
> 	},
> 	`map_index_keys`: {
> ```

is this a typo? this was previously called "max_index_keys".

---

*[pkg/sql/vars.go, line 103 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg7NAsVOroHPVOm1Od:-Kfg7NAsVOroHPVOm1Oe:b-mxrrxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4784ea221c39953fa645cd8be4a3717c58119c13/pkg/sql/vars.go#L103)):*
> ```Go
> 		Reset: func(p *planner) error {
> 			var zeroVal parser.Syntax
> 			p.session.Syntax = zeroVal
> ```

why not `p.session.Syntax = parser.Syntax(0)`? ditto throughout.

---

*[pkg/sql/parser/set.go, line 47 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg8HdMlHBQLM6bP_8R:-Kfg8HdMlHBQLM6bP_8S:b-mv4941) ([raw file](https://github.com/cockroachdb/cockroach/blob/b019d85e94e18454dd46db2a6bcd93e7636b7e7a/pkg/sql/parser/set.go#L47)):*
> ```Go
> // Format implements the NodeFormatter interface.
> func (node *Set) Format(buf *bytes.Buffer, f FmtFlags) {
> 	if node.SetMode == SetModeReset {
> ```

nit: slight preference for this being written as a switch-case.

---

*[pkg/sql/testdata/set, line 138 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg7myhuTbHy_RJojM9:-Kfg7myhuTbHy_RJojMA:bb6m372) ([raw file](https://github.com/cockroachdb/cockroach/blob/4784ea221c39953fa645cd8be4a3717c58119c13/pkg/sql/testdata/set#L138)):*
> ```
> SHOW APPLICATION_NAME
> ----
> application_name
> ```

what am I looking at here? empty string?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-Kfg8hzM12BfztAg4SIB:bfb2qpy)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Review status: 1 of 13 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[pkg/sql/vars.go, line 38 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg5AGbQ1wMCy8YTdNZ:-Kfjw8JQBwiOtmKRa_cn:bgcyp0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/vars.go#L38)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`ctx` should be the first argument (I think we might even lint for this)

also (nit): this can be written without names: `Set func(context.Context, *planner, []parser.TypedExpr) error` -- ditto below.
</blockquote></details>

Sure!

The reason I wrote it with names is because although this isn't an exported struct, if it were the parameter names would be a useful part of the godoc comment. I'm happy to remove though if you like?

---

*[pkg/sql/vars.go, line 223 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6ndclgUGydB8A1PD:-KfjwmIYl4f0V_s1NwZa:b-ps2dq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/vars.go#L223)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this a typo? this was previously called "max_index_keys".
</blockquote></details>

Whoops! Good catch, thanks! :smile: 

---

*[pkg/sql/vars.go, line 103 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg7NAsVOroHPVOm1Od:-KfgUIpmRuT123nz9eyl:b8s64fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4784ea221c39953fa645cd8be4a3717c58119c13/pkg/sql/vars.go#L103)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not `p.session.Syntax = parser.Syntax(0)`? ditto throughout.
</blockquote></details>

Good point! silly me

---

*[pkg/sql/testdata/set, line 138 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg7myhuTbHy_RJojM9:-KfjwbSCBZgys2sHex0o:bs6wme3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4784ea221c39953fa645cd8be4a3717c58119c13/pkg/sql/testdata/set#L138)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what am I looking at here? empty string?
</blockquote></details>

Yep, which is the default value for `APPLICATION_NAME` - is there a better way of calling this out in the test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Hey @tamird, I've now squashed my last two commits into one and either applied your feedback or responded in Reviewable.

Thanks again!

tamird commented:
Reviewed 4 of 15 files at r5, 2 of 12 files at r12, 10 of 10 files at r13.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KfkQWw_gVSD9vSEnzag:b-82oqa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you do the same origName thing here that is (was) done in show?
</blockquote></details>

Ping.

---

*[pkg/sql/vars.go, line 38 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg5AGbQ1wMCy8YTdNZ:-KfkQD4sJIR6oAV3Zazz:b-yzdiyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/vars.go#L38)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Sure!

The reason I wrote it with names is because although this isn't an exported struct, if it were the parameter names would be a useful part of the godoc comment. I'm happy to remove though if you like?
</blockquote></details>

Nah, keep them if you think they are useful.

---

*[pkg/sql/testdata/set, line 138 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg7myhuTbHy_RJojM9:-KfkPwXjH3DjzVuU280a:b-d4nvbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4784ea221c39953fa645cd8be4a3717c58119c13/pkg/sql/testdata/set#L138)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Yep, which is the default value for `APPLICATION_NAME` - is there a better way of calling this out in the test?
</blockquote></details>

I was thinking you'd use a variable with a non-empty default to call this out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Review status: 12 of 13 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KflKtS9BMs5YgcyNm-w:b-p4akzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Ping.
</blockquote></details>

Oh, sorry - maybe I misunderstood?

I've changed both instances to use the original-cased name (`origName`), is that not what you meant?

---

*[pkg/sql/testdata/set, line 138 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg7myhuTbHy_RJojM9:-KflKoEDnNhPF9bbLpKY:bi5uok8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4784ea221c39953fa645cd8be4a3717c58119c13/pkg/sql/testdata/set#L138)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I was thinking you'd use a variable with a non-empty default to call this out.
</blockquote></details>

Done :smiley: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KflLUB9WJK0DH7rVtIn:b2h31l4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Oh, sorry - maybe I misunderstood?

I've changed both instances to use the original-cased name (`origName`), is that not what you meant?
</blockquote></details>

This is still using `name`, not `origName`, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KflM-zeNNnLVM4wqtHj:b-ifv4ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

This is still using `name`, not `origName`, though.
</blockquote></details>

Are we definitely looking at the [same revision](https://github.com/boxofrad/cockroach/blob/24d6538cda2668c1a08aec15cba845bd084311f3/pkg/sql/set.go#L58)? or do you mean the other error messages (e.g. 'variable "foo" cannot be changed')

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KflO1bPwbh0nBqoVHJ2:b-c6lwq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Are we definitely looking at the [same revision](https://github.com/boxofrad/cockroach/blob/24d6538cda2668c1a08aec15cba845bd084311f3/pkg/sql/set.go#L58)? or do you mean the other error messages (e.g. 'variable "foo" cannot be changed')
</blockquote></details>

Yes, all the error messages should use the original casing (or else none of them should): https://github.com/boxofrad/cockroach/blob/24d6538cda2668c1a08aec15cba845bd084311f3/pkg/sql/set.go#L69:L76

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KflOJRKsMNAQYx18UX1:b-2h0e6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yes, all the error messages should use the original casing (or else none of them should): https://github.com/boxofrad/cockroach/blob/24d6538cda2668c1a08aec15cba845bd084311f3/pkg/sql/set.go#L69:L76
</blockquote></details>

In fact, you might consider inlining `name := strings.ToLower(origName)` into `v, ok := varGen[strings.ToLower(origName)]` to avoid this class of mistake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Review status: 11 of 13 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/set.go, line 63 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kfg6bzMM5otRVrvtS1B:-KflQgvfQyYytwS5o_0L:bcfgcw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d58d767319efedfedd988586af25112856ec790b/pkg/sql/set.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

In fact, you might consider inlining `name := strings.ToLower(origName)` into `v, ok := varGen[strings.ToLower(origName)]` to avoid this class of mistake.
</blockquote></details>

Gotcha, sorry about the misunderstanding!

Have inlined the `ToLower` as suggested :smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

@knz and @jordanlewis should look over this as well.

---

Reviewed 2 of 2 files at r15.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/show.go, line 147 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-KflWMgClJ4yO8Epzgw6:-KflWMgEb1XuTtatVMQw:bf2dasg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2ab4349af1de189eb5d8f7d90669333243b40a3/pkg/sql/show.go#L147)):*
> ```Go
> 	var columns ResultColumns
> 
> 	switch name {
> ```

super nit: like the other site, you could inline origName into the map lookup if you rewrite the switch as:

```
	if strings.EqualFold(name, `all`) {
		columns = ResultColumns{
			{Name: "Variable", Typ: parser.TypeString},
			{Name: "Value", Typ: parser.TypeString},
		}
	} else {
		if _, ok := varGen[strings.ToLower(name)]; !ok {
			....
		}
		...
	}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-KflX1PEblGyQgMamUK4:b-up1tc9)*
<!-- Sent from Reviewable.io -->

knz commented:
LGTM with some nits

---

Reviewed 2 of 12 files at r12, 7 of 10 files at r13, 1 of 1 files at r14, 2 of 2 files at r15.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/set.go, line 50 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kflyq2a0iWcyNxid-PJ:-Kflyq2a0iWcyNxid-PK:bebpvge) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1bb480d24ff747b2297c9333f9f1286bd5dda04/pkg/sql/parser/set.go#L50)):*
> ```Go
> 	case SetModeAssign:
> 		buf.WriteString("SET ")
> 		if node.Name != nil {
> ```

Could you briefly run a test whether `node.Name` is ever set to `nil`. It seems to me that `SET` always has a name here. If that is true, then you can drop the `if`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-Kflz8DLacPXpC7kZc01:b77d3f7)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/show.go, line 147 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-KflWMgClJ4yO8Epzgw6:-Kfm70A_6BR7jHwPdp05:b-vmsteq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2ab4349af1de189eb5d8f7d90669333243b40a3/pkg/sql/show.go#L147)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

super nit: like the other site, you could inline origName into the map lookup if you rewrite the switch as:

```
	if strings.EqualFold(name, `all`) {
		columns = ResultColumns{
			{Name: "Variable", Typ: parser.TypeString},
			{Name: "Value", Typ: parser.TypeString},
		}
	} else {
		if _, ok := varGen[strings.ToLower(name)]; !ok {
			....
		}
		...
	}
```
</blockquote></details>

Good point, although I'll likely need to introduce a couple of local variables to store the result of `EqualFold` and the found member of the `varGen` map, as there's also an anonymous function below (`delayedNode.constructor`) which also performs this switch and map lookup.

Mind if I give this one a miss?

---

*[pkg/sql/parser/set.go, line 50 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kflyq2a0iWcyNxid-PJ:-Kfm5SfwnmQdgVS7pvmq:bwzqrfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1bb480d24ff747b2297c9333f9f1286bd5dda04/pkg/sql/parser/set.go#L50)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Could you briefly run a test whether `node.Name` is ever set to `nil`. It seems to me that `SET` always has a name here. If that is true, then you can drop the `if`.
</blockquote></details>

Took a quick look at the git blame, it seems this conditional was introduced in `d2cbf0e`to handle the `SET ROW` syntax.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/set.go, line 50 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-Kflyq2a0iWcyNxid-PJ:-KfmArNsTRTlXhNPBMyA:bmhtocv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1bb480d24ff747b2297c9333f9f1286bd5dda04/pkg/sql/parser/set.go#L50)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Took a quick look at the git blame, it seems this conditional was introduced in `d2cbf0e`to handle the `SET ROW` syntax.
</blockquote></details>

D'oh! That's mine, and I had forgotten :)
Thanks for checking though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/show.go, line 147 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-KflWMgClJ4yO8Epzgw6:-KfmDxtxadp94M2tMImA:bk0ezuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2ab4349af1de189eb5d8f7d90669333243b40a3/pkg/sql/show.go#L147)):*
<details><summary><i>Previously, boxofrad (Daniel Upton) wrote…</i></summary><blockquote>

Good point, although I'll likely need to introduce a couple of local variables to store the result of `EqualFold` and the found member of the `varGen` map, as there's also an anonymous function below (`delayedNode.constructor`) which also performs this switch and map lookup.

Mind if I give this one a miss?
</blockquote></details>

Ah, makes sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
This is good to go; can you please rebase (there's a merge conflict right now) and then we can merge this PR.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-Kg5TzoJoVdMEmyRyB1k:b-3ou42z)*
<!-- Sent from Reviewable.io -->

boxofrad commented:
Rebased and ready to merge :smile:

---

Review status: 1 of 13 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105#-:-KgBJjOXK-_pRHie6rlS:b-49j6ra)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 14 of 14 files at r16, 10 of 10 files at r17.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
You need to rebase and re-generate `sql.go` once more, sorry for the inconvenience. We had a couple of "hot" PRs in the front-end lately.
If this is costing too much of your time please let me know and I take the last phase over from you.

boxofrad commented:
No worries, have rebased & regenerated.

tamird commented:
Looks like there's still a conflict.

boxofrad commented:
Sorry for my delay in getting back, it looks like master has diverged quite a bit now (noticed the new `requireExplicitTxn` member in the `varGen` map).

I may not have time to rebase this branch again (in the next couple of days at least) would anybody else be able to help? 😃

dt commented:
@boxofrad Thanks for the heads up! I give this a rebase. I don't seem to be able to push directly to the PR branch (is `Allow edits from maintainers`  enabled for this PR?), so I've pushed the rebased refs to  boxofrad-reset-default branch on origin for now if you want to push them to your branch.

boxofrad commented:
Wow thanks @dt, really appreciate you stepping in!

I've pushed your changes to my branch, and ticked the magic checkbox so you can push to my branch should you need to in the future.

tamird commented:
Reviewed 3 of 13 files at r18, 10 of 10 files at r19.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 13 files at r18, 10 of 10 files at r19.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14105)*
<!-- Sent from Reviewable.io -->

knz commented:
Thank you @dt for your stewardship. Thanks @boxofrad for your contribution!
